### PR TITLE
Fix POW not being returned when compute leafs is on

### LINKED ISSUE #66: Fix gpu-post not returning a found PoW when both ComputeLeafs and ComputePow are set
At the moment a found PoW will never be returned if `ComputeLeafs` is also set while running `scryptPositions`.

### PR DESCRIPTION
**Pull request by fasmat**
Fixes #66 

There seems to be a bug where if `computeLeafs` is set to true, any valid PoW that is found is discarded instead of returned to the caller.

### CODE REVIEW

File: src/scrypt-jane/scrypt-jane.cpp [hunk 1]
```diff
@@ -499,152 +499,126 @@ static int cpu_scrypt_positions(
...
-							if (!computeLeafs) {
-								break;
-							}
-						}
-					}
-				}
-			}
-			else {
-				uint32_t i;
+	for (; n <= end_position && !g_spacemesh_api_abort_flag; n++) {
```
Comment (poszu) [inline src/scrypt-jane/scrypt-jane.cpp:533]: A `for (; <condition>; ) {...}`  behaves differently than `do {...} while (<condition>)`. The latter is guaranteed to run at least once. Can you motivate this change?

File: src/scrypt-jane/scrypt-jane.cpp [hunk 2]
```diff
@@ -499,152 +499,126 @@ static int cpu_scrypt_positions(
...
+			if (!computeLeafs) {
+				break;
 			}
-#if 0
-			if (8 == hash_len_bits) {
-				*output++ = label;
+		}
+		if (!computeLeafs) {
+			continue;
+		}
```
Comment (poszu) [inline src/scrypt-jane/scrypt-jane.cpp:563]: This whole loop is terribly convoluted. Adding `continue` doesn't help to understand the flow.

**Review by poszu — CHANGES_REQUESTED** [review]
Could you write a unit test that would cover the faulty case?

File: src/scrypt-jane/scrypt-jane.cpp (hunk 1 above)
Comment (fasmat) [reply src/scrypt-jane/scrypt-jane.cpp:533]: In this case the for is also guaranteed to run at least once (Endposition >= Startposition and the abort flag is unlikely to be set immediately when this function starts).

I think it makes the code easier to read and I can add `n++` to the for statement instead of having it in multiple places in the body.

File: src/scrypt-jane/scrypt-jane.cpp (hunk 2 above)
Comment (fasmat) [reply src/scrypt-jane/scrypt-jane.cpp:563]: I inverted some ifs to make the for loop easier to understand, that's where the `continue` comes from.

Comment (fasmat) [reply src/scrypt-jane/scrypt-jane.cpp:563]: It helps not using the diff view but looking at the two versions separately to see the changes.

**Comment by fasmat** [comment]
Will do today.

File: .gitignore [hunk 3]
```diff
@@ -54,3 +54,9 @@ _UpgradeReport_Files/
 Backup*/
 UpgradeLog*.XML
 UpgradeLog*.htm
+
+# CMake files
+CMakeCache.txt
+CMakeFiles
+Makefile
+cmake_install.cmake
```
Comment (poszu) [inline .gitignore]: Cmake should rather create its files in a directory like `build/`.

File: test/test.cpp [hunk 4]
```diff
@@ -462,7 +462,114 @@ int do_test_pow(uint64_t aStartPos, int aLabelsCount, unsigned aDiff, unsigned a
...
+							printf("\n");
+							printf("salt: ");
+							print_hex32(s_salt);
+							printf("\n");
+							printf("D: ");
+							print_hex32(D);
+							printf("\n");
+							printf("H: ");
+							print_hex32(hash);
+							printf("\n");
```
Comment (poszu) [inline test/test.cpp]: Why print so much in tests?

File: test/test.cpp [hunk 5]
```diff
@@ -462,7 +462,114 @@ int do_test_pow(uint64_t aStartPos, int aLabelsCount, unsigned aDiff, unsigned a
...
+			}
+
+			return 0;
+		}
+	}
+
+	return 1;
+}
+
+/* test PoW */
```
Comment (poszu) [inline test/test.cpp]: This comment doesn't have much value and it's the same as the test above. Could you elaborate more on what this test is about?

File: test/test.cpp [hunk 6]
```diff
@@ -462,7 +462,114 @@ int do_test_pow(uint64_t aStartPos, int aLabelsCount, unsigned aDiff, unsigned a
...
+			}
+			for (int i = 0; i < providersCount; i++) {
+				if (providersCount == 1 || providers[i].compute_api != COMPUTE_API_CLASS_CPU)
+				{
+					if (-1 != aProviderId && aProviderId != providers[i].id) {
+						continue;
+					}
+
+					const uint64_t labelsBufferSize = (uint64_t(aLabelsCount) * uint64_t(labelSize) + 7ull) / 8ull;
+					uint8_t *out = (uint8_t *)malloc(labelsBufferSize);
```
Comment (poszu) [inline test/test.cpp:558]: Given that it's a test that will exit soon, it's not a big deal - I think that `out` is not freed.

File: .gitignore (hunk 3 above)
Comment (fasmat) [reply .gitignore]: How would you do that? I can specify where compiled binaries are written to (already to `build`), but intermediary files (like `Makefile` and `CMakeCache.txt`) are written to `.` by `cmake`.

Otherwise the build also needs to change directories between steps:

```bash
cmake
cd build
make
```

File: test/test.cpp (hunk 4 above)
Comment (fasmat) [reply test/test.cpp]: I just copied an existing test and modified it to test for the specific change I made.

I can remove those print statement for the new test if you want.

File: test/test.cpp (hunk 5 above)
Comment (fasmat) [reply test/test.cpp]: updated

File: src/scrypt-jane/scrypt-jane.cpp (hunk 2 above)
Comment (poszu) [reply src/scrypt-jane/scrypt-jane.cpp:563]: I understand where it comes from. I meant it's hard to follow this for loop in general. There are too many branches and `continue`s.

**Review by poszu — APPROVED** [review]
